transport/socks5/stream_dialer.go

```diff
@@ -206,32 +198,47 @@ func (c *StreamDialer) DialStream(ctx context.Context, remoteAddr string) (trans
 		return nil, ReplyCode(buffer[1])
 	}
 
-	// 4. Read address and length
-	var bndAddrLen int
-	switch buffer[3] {
-	case addrTypeIPv4:
-		bndAddrLen = 4
-	case addrTypeIPv6:
-		bndAddrLen = 16
-	case addrTypeDomainName:
-		// buffer[8]: length of the domain name
-		_, err := io.ReadFull(proxyConn, buffer[:1])
-		if err != nil {
-			return nil, fmt.Errorf("failed to read address length in connect response: %w", err)
-		}
-		bndAddrLen = int(buffer[0])
-	default:
-		return nil, fmt.Errorf("invalid address type %v", buffer[3])
-	}
-	// 5. Reads the bound address and port, but we currently ignore them.
-	// TODO(fortuna): Should we expose the remote bound address as the net.Conn.LocalAddr()?
-	if _, err := io.ReadFull(proxyConn, buffer[:bndAddrLen]); err != nil {
+	// 4. Read BND.ADDR.
+	bindAddr, err := readAddr(conn)
+	if err != nil {
 		return nil, fmt.Errorf("failed to read bound address: %w", err)
 	}
-	// We read but ignore the remote bound port number: BND.PORT
-	if _, err = io.ReadFull(proxyConn, buffer[:2]); err != nil {
-		return nil, fmt.Errorf("failed to read bound port: %w", err)
+
+	return bindAddr, nil
+}
+
+// connectAndRequest manages the connection lifecycle and delegates the SOCKS5 communication to the request function.
+func (c *Client) connectAndRequest(ctx context.Context, cmd byte, dstAddr string) (transport.StreamConn, *address, error) {
+	proxyConn, err := c.se.ConnectStream(ctx)
+	if err != nil {
+		return nil, nil, fmt.Errorf("could not connect to SOCKS5 proxy: %w", err)
+	}
+
+	// Close the connection if an error occurs in any subsequent steps
+	// to avoid keeping the connection open.
+	defer func(conn transport.StreamConn) {
+		if conn != nil && err != nil {
```

Comment:
conn is never nil, because you are passing it as a parameter on successful ConnectStream.

I think it's cleaner if you just call close in the case of request error below, given the logic is simple.